pageserver/src/tenant/timeline.rs

```diff
@@ -118,6 +121,80 @@ impl PartialOrd for Hole {
     }
 }
 
+pub struct LayerMapping(HashMap<PersistentLayerKey, Arc<dyn PersistentLayer>>);
+
+impl LayerMapping {
+    fn get_from_desc(&self, desc: &PersistentLayerDesc) -> Arc<dyn PersistentLayer> {
+        self.0.get(&desc.key()).expect("not found").clone()
+    }
+
+    pub(crate) fn insert(&mut self, layer: Arc<dyn PersistentLayer>) {
+        self.0.insert(layer.layer_desc().key(), layer);
+    }
+
+    pub(crate) fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    pub(crate) fn remove(&mut self, layer: Arc<dyn PersistentLayer>) {
+        self.0.remove(&layer.layer_desc().key());
+    }
+
+    pub(crate) fn replace_and_verify(
+        &mut self,
+        expected: Arc<dyn PersistentLayer>,
+        new: Arc<dyn PersistentLayer>,
+    ) -> Result<()> {
+        use super::layer_map::LayerKey;
+        let key = LayerKey::from(&*expected);
+        let other = LayerKey::from(&*new);
+
+        let expected_l0 = LayerMap::is_l0(expected.layer_desc());
+        let new_l0 = LayerMap::is_l0(new.layer_desc());
+
+        fail::fail_point!("layermap-replace-notfound", |_| anyhow::bail!(
+            "replacing downloaded layer into layermap failed because layer was not found"
```

Comment:
Typically our failpoint errors have the exact same message as the failpoint name. Any reason why that isn't the case here?

Reply:
The original code in `layer_map` is

```
fail::fail_point!("layermap-replace-notfound", |_| Ok(Replacement::NotFound));
```

I can change it to be the exactly same error message as the failpoint name.

Reply:
`Replacement::NotFound` was used originally because it would had triggered the busy-loop before the #3613 (unable to reproduce the real situation).